plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,251 @@
+package snmp_trap
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+
+	"github.com/soniah/gosnmp"
+)
+
+type handler func(*gosnmp.SnmpPacket, *net.UDPAddr)
+type execer func(string, ...string) ([]byte, error)
+
+type mibEntry struct {
+	mibName string
+	oidText string
+}
+
+type SnmpTrap struct {
+	ServiceAddress string `toml:"service_address"`
+
+	acc      telegraf.Accumulator
+	listener *gosnmp.TrapListener
+	timeFunc func() time.Time
+	errCh    chan error
+
+	makeHandlerWrapper func(handler) handler
+
+	Log telegraf.Logger `toml:"-"`
+
+	cacheLock sync.Mutex
+	cache     map[string]mibEntry
+
+	execCmd execer
+}
+
+var sampleConfig = `
+  ## Transport, local address, and port to listen on.  Transport must
+  ## be "udp://".  Omit local address to listen on all interfaces.
+  ##   example: "udp://127.0.0.1:1234"
+  # service_address = udp://:162
+`
+
+func (s *SnmpTrap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *SnmpTrap) Description() string {
+	return "Receive SNMP traps"
+}
+
+func (s *SnmpTrap) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func init() {
+	inputs.Add("snmp_trap", func() telegraf.Input {
+		return &SnmpTrap{
+			timeFunc:       time.Now,
+			ServiceAddress: "udp://:162",
+		}
+	})
+}
+
+func realExecCmd(arg0 string, args ...string) ([]byte, error) {
+	return exec.Command(arg0, args...).Output()
+}
+
+func (s *SnmpTrap) Init() error {
+	s.cache = map[string]mibEntry{}
+	s.execCmd = realExecCmd
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+
+	// wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	split := strings.SplitN(s.ServiceAddress, "://", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("invalid service address: %s", s.ServiceAddress)
+	}
+
+	protocol := split[0]
+	addr := split[1]
+
+	// gosnmp.TrapListener currentl supports udp only.  For forward
+	// compatibility, require udp in the service address
+	if protocol != "udp" {
+		return fmt.Errorf("unknown protocol '%s' in '%s'", protocol, s.ServiceAddress)
+	}
+
+	// If (*TrapListener).Listen immediately returns an error we need
+	// to return it from this function.  Use a channel to get it here
+	// from the goroutine.  Buffer one in case Listen returns after
+	// Listening but before our Close is called.
+	s.errCh = make(chan error, 1)
+	go func() {
+		s.errCh <- s.listener.Listen(addr)
+	}()
+
+	select {
+	case <-s.listener.Listening():
+		s.Log.Infof("Listening on %s", s.ServiceAddress)
+	case err := <-s.errCh:
+		return err
+	}
+
+	return nil
+}
+
+func (s *SnmpTrap) Stop() {
+	s.listener.Close()
+	_ = <-s.errCh
```

Comment:
Not required, will be garbage collected, this would allow you to make the channel a local variable.

Reply:
Instead of making s.errCh local to Start() I left it in the SnmpTrap struct and changed Stop() to log if there's an error.